content/exchange/artifacts/Server.Hunt.Comparison.yaml

```diff
@@ -0,0 +1,57 @@
+name: Server.Hunt.Comparison
+author: Denis Kiffer
+
+description: |
+   This artifact is used to compare other artifacts from two different hunts. The basic idea is that a baseline (Hunt 1) is created from selected artifacts before an attack. A second hunt (Hunt 2) can then be carried out after the attack using the same artifacts. Now, using this script, artifacts from both hunts can be compared. This allows legitimate activities to be filtered out and makes it easier to identify malicious activities in Hunt2.
+   
+   Furthermore, when comparing artifacts, it is necessary to select columns (here the identifying_column) that should be used for the comparison in both artifacts, since a comparison of complete data sets leads to errors. Because many artifacts contain timestamps that update. The use of values such as hashes therefore makes sense.
+   
+   For example, the following artifacts and their identifying columns can be used to compare to a baseline:
+   
+   - Windows.System.Pslist - Hash
+   - Windows.Forensics.Prefetch - Hash
+   - Windows.Sysinternals.Autoruns - SHA-256
+   - Windows.Sys.AllUsers - Name
+   
+reference:
+  - https://www.sans.org/white-papers/37192/
+
+type: SERVER
+
+parameters:
+  - name: Baseline_Hunt_1_id
+    description: id of the hunt that was executed before an attack
+  - name: Hunt_2_id
+    description: id of hunt that was executed during or after an attack and is supposed to be compared to the Baseline_Hunt_1
+  - name: Artifact_name
+    description: This is the artefact that should be compared and which was executed in the Baseline_Hunt_1 and Hunt_2
+  - name: Identifying_column
+    description: column of the selected artefact that should be compared
+
+sources:
+  - query: |
+     --Select baseline hunt
+     Let Baseline = SELECT *,"Baseline" AS Sourcehunt FROM hunt_results(
+       artifact=Artifact_name,
+       hunt_id=Baseline_Hunt_1_id) 
+
+     --Select second hunt to compare    
+     Let Hunt2 = SELECT *, "Hunt2" AS Sourcehunt FROM hunt_results(
+       artifact=Artifact_name,
+       hunt_id=Hunt_2_id)
+
+     --Get List of systems to compare to each other    
+     Let Systems = SELECT Fqdn FROM Baseline GROUP BY Fqdn
+
+     --Fuse both hunts 
+     Let Hunts_fused = SELECT * FROM chain(
+       a={SELECT * FROM Hunt2},
+       b={SELECT * FROM Baseline},async=TRUE)
+
+     --Loop through each system and count how many appearances of the value from the selected identifying_column there are           
+     Let CountallRows = SELECT * FROM foreach(
+       row=Systems.Fqdn,
+       query={SELECT *, count() AS TotalCount FROM Hunts_fused WHERE Fqdn=_value GROUP BY get(member=Identifying_column)})
+
+     --Show all lines where the value in the identifying_column only appears once
+     SELECT * FROM CountallRows WHERE TotalCount=1
```

Comment:
Im not sure this really works - in practice there may be a lot of instances of the same values.

Here is a counter example - Say I run a hunt for `Windows.System.Pslist` and look at the hash (your example in the doc). This will show every instance of notepad running on the same system - so I should expect this count to be much larger than 1 even with  a single hunt.

Additionally this query counts all the same signal across all clients so if two or more clients are running a binary its count will be larger than 1 - even if it is new than the baseline.

Reply:
You're probably right with your first argument. If an executable is run multiple times on a client, the number should be higher. But in my experience, malware usually only runs in one instance on an infected client in reality.

But if I understand your second statement correctly, that the number will be higher than 1 because many clients are running the same executable on a network, I would disagree with that. Because each client is only compared to itself (from the past) because the foreach function iterates through each Fqdn.

Reply:
I actually just wrote my master's thesis on the topic and tried this approach on a previously compromised multi-system virtual network and it worked quite well.